Travis error:

```
Skipping hack/../hack/verify-all.sh

Boilerplate header is wrong for: /home/travis/gopath/src/k8s.io/contrib/ansible/library/oc_atomic_container.py

!!! 'golint' problems: 

/home/travis/gopath/src/k8s.io/contrib/kubeform/pkg/provider/provider.go:428:2: redundant if ...; err != nil check, just return error instead.
```

--- a/ansible/inventory/group_vars/all.yml
+++ b/ansible/inventory/group_vars/all.yml
@@ -134,3 +134,12 @@ etcd_client_port: '2379'
 # when the scheme vars above are set to "https" you need to set these to true!
 flannel_etcd_use_certs: true
 apiserver_etcd_use_certs: true
+
+kube_use_node_system_container: false
+kube_use_master_system_container: false
+kube_use_etcd_system_container: false
+kube_use_flannel_system_container: false
+
+system_images_registry: "candidate-registry.fedoraproject.org"
+system_images_registry_namespace: "f26"
+system_images_tag: "latest"

--- a/ansible/library/oc_atomic_container.py
+++ b/ansible/library/oc_atomic_container.py
@@ -1,0 +1,216 @@
+#!/usr/bin/env python
+# pylint: disable=missing-docstring
+# flake8: noqa: T001
+#     ___ ___ _  _ ___ ___    _ _____ ___ ___
+#    / __| __| \| | __| _ \  /_\_   _| __|   \
+#   | (_ | _|| .` | _||   / / _ \| | | _|| |) |
+#    \___|___|_|\_|___|_|_\/_/_\_\_|_|___|___/_ _____
+#   |   \ / _ \  | \| |/ _ \_   _| | __|   \_ _|_   _|
+#   | |) | (_) | | .` | (_) || |   | _|| |) | |  | |
+#   |___/ \___/  |_|\_|\___/ |_|   |___|___/___| |_|
+#
+# Copyright 2016 Red Hat, Inc. and/or its affiliates
+# and other contributors as indicated by the @author tags.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# -*- -*- -*- Begin included fragment: doc/atomic_container -*- -*- -*-
+
+DOCUMENTATION = '''
+---
+module: oc_atomic_container
+short_description: Manage the container images on the atomic host platform
+description:
+    - Manage the container images on the atomic host platform
+    - Allows to execute the commands on the container images
+requirements:
+  - atomic
+  - "python >= 2.6"
+options:
+    name:
+        description:
+          - Name of the container
+        required: True
+        default: null
+    image:
+        description:
+          - The image to use to install the container
+        required: True
+        default: null
+    state:
+        description:
+          - State of the container
+        required: True
+        choices: ["latest", "absent", "latest", "rollback"]
+        default: "latest"
+    values:
+        description:
+          - Values for the installation of the container
+        required: False
+        default: None
+'''
+
+# -*- -*- -*- End included fragment: doc/atomic_container -*- -*- -*-
+
+# -*- -*- -*- Begin included fragment: ansible/oc_atomic_container.py -*- -*- -*-
+
+# pylint: disable=wrong-import-position,too-many-branches,invalid-name,no-name-in-module, import-error
+import json
+
+from distutils.version import StrictVersion
+
+from ansible.module_utils.basic import AnsibleModule
+
+
+def _install(module, container, image, values_list):
+    ''' install a container using atomic CLI.  values_list is the list of --set arguments.
+    container is the name given to the container.  image is the image to use for the installation. '''
+    # NOTE: system-package=no is hardcoded. This should be changed to an option in the future.
+    args = ['atomic', 'install', '--system', '--system-package=no',
+            '--name=%s' % container] + values_list + [image]
+    rc, out, err = module.run_command(args, check_rc=False)
+    if rc != 0:
+        return rc, out, err, False
+    else:
+        changed = "Extracting" in out or "Copying blob" in out
+        return rc, out, err, changed
+
+def _uninstall(module, name):
+    ''' uninstall an atomic container by its name. '''
+    args = ['atomic', 'uninstall', name]
+    rc, out, err = module.run_command(args, check_rc=False)
+    return rc, out, err, False
+
+
+def do_install(module, container, image, values_list):
+    ''' install a container and exit the module. '''
+    rc, out, err, changed = _install(module, container, image, values_list)
+    if rc != 0:
+        module.fail_json(rc=rc, msg=err)
+    else:
+        module.exit_json(msg=out, changed=changed)
+
+
+def do_uninstall(module, name):
+    ''' uninstall a container and exit the module. '''
+    rc, out, err, changed = _uninstall(module, name)
+    if rc != 0:
+        module.fail_json(rc=rc, msg=err)
+    module.exit_json(msg=out, changed=changed)
+
+
+def do_update(module, container, old_image, image, values_list):
+    ''' update a container and exit the module.  If the container uses a different
+    image than the current installed one, then first uninstall the old one '''
+
+    # the image we want is different than the installed one
+    if old_image != image:
+        rc, out, err, _ = _uninstall(module, container)
+        if rc != 0:
+            module.fail_json(rc=rc, msg=err)
+        return do_install(module, container, image, values_list)
+
+    # if the image didn't change, use "atomic containers update"
+    args = ['atomic', 'containers', 'update'] + values_list + [container]
+    rc, out, err = module.run_command(args, check_rc=False)
+    if rc != 0:
+        module.fail_json(rc=rc, msg=err)
+    else:
+        changed = "Extracting" in out or "Copying blob" in out
+        module.exit_json(msg=out, changed=changed)
+
+
+def do_rollback(module, name):
+    ''' move to the previous deployment of the container, if present, and exit the module. '''
+    args = ['atomic', 'containers', 'rollback', name]
+    rc, out, err = module.run_command(args, check_rc=False)
+    if rc != 0:
+        module.fail_json(rc=rc, msg=err)
+    else:
+        changed = "Rolling back" in out
+        module.exit_json(msg=out, changed=changed)
+
+
+def core(module):
+    ''' entrypoint for the module. '''
+    name = module.params['name']
+    image = module.params['image']
+    values = module.params['values']
+    state = module.params['state']
+
+    module.run_command_environ_update = dict(LANG='C', LC_ALL='C', LC_MESSAGES='C')
+    out = {}
+    err = {}
+    rc = 0
+
+    values_list = ["--set=%s" % x for x in values] if values else []
+
+    args = ['atomic', 'containers', 'list', '--json', '--all', '-f', 'container=%s' % name]
+    rc, out, err = module.run_command(args, check_rc=False)
+    if rc != 0:
+        module.fail_json(rc=rc, msg=err)
+        return
+
+    # NOTE: "or '[]' is a workaround until atomic containers list --json
+    # provides an empty list when no containers are present.
+    containers = json.loads(out or '[]')
+    present = len(containers) > 0
+    old_image = containers[0]["image_name"] if present else None
+
+    if state == 'present' and present:
+        module.exit_json(msg=out, changed=False)
+    elif (state in ['latest', 'present']) and not present:
+        do_install(module, name, image, values_list)
+    elif state == 'latest':
+        do_update(module, name, old_image, image, values_list)
+    elif state == 'absent':
+        if not present:
+            module.exit_json(msg="", changed=False)
+        else:
+            do_uninstall(module, name)
+    elif state == 'rollback':
+        do_rollback(module, name)
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(default=None, required=True),
+            image=dict(default=None, required=True),
+            state=dict(default='latest', choices=['present', 'absent', 'latest', 'rollback']),
+            values=dict(type='list', default=[]),
+            ),
+        )
+
+    # Verify that the platform supports atomic command
+    rc, version_out, err = module.run_command('rpm -q --queryformat "%{VERSION}\n" atomic', check_rc=False)
+    if rc != 0:
+        module.fail_json(msg="Error in running atomic command", err=err)
+    # This module requires atomic version 1.17.2 or later
+    atomic_version = StrictVersion(version_out.replace('\n', ''))
+    if atomic_version < StrictVersion('1.17.2'):
+        module.fail_json(
+            msg="atomic version 1.17.2+ is required",
+            err=str(atomic_version))
+
+    try:
+        core(module)
+    except Exception as e:  # pylint: disable=broad-except
+        module.fail_json(msg=str(e))
+
+
+if __name__ == '__main__':
+    main()
+
+# -*- -*- -*- End included fragment: ansible/oc_atomic_container.py -*- -*- -*-

--- a/ansible/library/oc_atomic_container.py
+++ b/ansible/library/oc_atomic_container.py
@@ -1,31 +1,18 @@
 #!/usr/bin/env python
-# pylint: disable=missing-docstring
-# flake8: noqa: T001
-#     ___ ___ _  _ ___ ___    _ _____ ___ ___
-#    / __| __| \| | __| _ \  /_\_   _| __|   \
-#   | (_ | _|| .` | _||   / / _ \| | | _|| |) |
-#    \___|___|_|\_|___|_|_\/_/_\_\_|_|___|___/_ _____
-#   |   \ / _ \  | \| |/ _ \_   _| | __|   \_ _|_   _|
-#   | |) | (_) | | .` | (_) || |   | _|| |) | |  | |
-#   |___/ \___/  |_|\_|\___/ |_|   |___|___/___| |_|
-#
-# Copyright 2016 Red Hat, Inc. and/or its affiliates
-# and other contributors as indicated by the @author tags.
+
+# Copyright 2017 The Kubernetes Authors.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
 # you may not use this file except in compliance with the License.
 # You may obtain a copy of the License at
 #
-#    http://www.apache.org/licenses/LICENSE-2.0
+#     http://www.apache.org/licenses/LICENSE-2.0
 #
 # Unless required by applicable law or agreed to in writing, software
 # distributed under the License is distributed on an "AS IS" BASIS,
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-#
-
-# -*- -*- -*- Begin included fragment: doc/atomic_container -*- -*- -*-
 
 DOCUMENTATION = '''
 ---

--- a/ansible/roles/common/tasks/main.yml
+++ b/ansible/roles/common/tasks/main.yml
@@ -25,6 +25,12 @@
     bin_dir: "/opt/bin"
   when: is_coreos
 
+- name: Set the bin directory path for system containers
+  set_fact:
+    bin_dir: "/usr/local/bin"
+  when: kube_use_master_system_container or kube_use_etcd_system_container
+
+
 - name: Create the directory used to store binaries
   file: path={{ bin_dir }} state=directory
 

--- a/ansible/roles/etcd/defaults/main.yaml
+++ b/ansible/roles/etcd/defaults/main.yaml
@@ -60,3 +60,5 @@ etcd_download_url: "{{ etcd_download_url_base }}/etcd-v{{ etcd_version }}-linux-
 etcd_enable: true
 
 etcd_is_proxy: false
+
+etcd_system_image: "{{ system_images_registry }}/{{ system_images_registry_namespace }}/etcd:{{system_images_tag }}"

--- a/ansible/roles/etcd/tasks/install.yml
+++ b/ansible/roles/etcd/tasks/install.yml
@@ -58,6 +58,10 @@
         state: latest
   when: etcd_source_type == "package-manager"
 
+- name: Install Etcd system container
+  include: system_container.yml
+  when: kube_use_etcd_system_container 
+
 - name: Set the etcd_modified fact to true
   set_fact:
     etcd_modified: true

--- a/ansible/roles/etcd/tasks/system_container.yml
+++ b/ansible/roles/etcd/tasks/system_container.yml
@@ -1,0 +1,10 @@
+- name: Pull etcd system container
+  command: atomic pull --storage=ostree {{ etcd_system_image }}
+  register: pull_result
+  changed_when: "'Pulling layer' in pull_result.stdout"
+
+- name: Install or Update Etcd system container package
+  oc_atomic_container:
+    name: etcd
+    image: "{{ etcd_system_image }}"
+    state: latest

--- a/ansible/roles/flannel/defaults/main.yaml
+++ b/ansible/roles/flannel/defaults/main.yaml
@@ -25,3 +25,5 @@ flannel_etcd_cert_file: "{{ flannel_etcd_certs_dir }}/client.crt"
 flannel_etcd_key_file: "{{ flannel_etcd_certs_dir }}/client.key"
 
 #flannel_options: --etcd-prefix=/coreos.com/network
+
+flannel_system_image: "{{ system_images_registry }}/{{ system_images_registry_namespace }}/flannel:{{ system_images_tag }}"

--- a/ansible/roles/flannel/tasks/client.yml
+++ b/ansible/roles/flannel/tasks/client.yml
@@ -13,6 +13,10 @@
         state: latest
   when: not is_atomic and flannel_source_type != "github-release"
 
+- name: Install flannel system container
+  include: system_container.yml
+  when: kube_use_flannel_system_container
+
 - set_fact:
     flannel_use_upstart: false
 

--- a/ansible/roles/flannel/tasks/config.yml
+++ b/ansible/roles/flannel/tasks/config.yml
@@ -15,7 +15,7 @@
   delegate_to: "{{ groups['etcd'][0] }}"
 
 - name: Load the flannel config file into etcd
-  shell: "/usr/bin/etcdctl {% if etcd_url_scheme is defined and etcd_url_scheme == 'https' %} --cert-file={{ flannel_etcd_cert_file }} --key-file={{ flannel_etcd_key_file }} --ca-file={{ flannel_etcd_ca_file }} {% endif %}--no-sync --peers={{ peers }} set {{ conf_loc }} < {{ conf_file }}"
+  shell: "{{ bin_dir }}/etcdctl {% if etcd_url_scheme is defined and etcd_url_scheme == 'https' %} --cert-file={{ flannel_etcd_cert_file }} --key-file={{ flannel_etcd_key_file }} --ca-file={{ flannel_etcd_ca_file }} {% endif %}--no-sync --peers={{ peers }} set {{ conf_loc }} < {{ conf_file }}"
   run_once: true
   delegate_to: "{{ groups['etcd'][0] }}"
 

--- a/ansible/roles/flannel/tasks/etcd-certs.yml
+++ b/ansible/roles/flannel/tasks/etcd-certs.yml
@@ -22,3 +22,6 @@
   with_together:
     - [ "{{ flannel_etcd_cert_file }}", "{{ flannel_etcd_ca_file }}", "{{ flannel_etcd_key_file }}" ]
     - [ "{{ etcd_client_cert }}", "{{ etcd_ca_cert }}", "{{ etcd_client_key }}" ]
+
+- set_fact:
+    flannel_options: "-etcd-cafile {{ flannel_etcd_ca_file }} -etcd-certfile {{ flannel_etcd_cert_file }} -etcd-keyfile {{ flannel_etcd_key_file }}"

--- a/ansible/roles/flannel/tasks/system_container.yml
+++ b/ansible/roles/flannel/tasks/system_container.yml
@@ -1,0 +1,10 @@
+- name: Pull flannel system container
+  command: atomic pull --storage=ostree {{ flannel_system_image }}
+  register: pull_result
+  changed_when: "'Pulling layer' in pull_result.stdout"
+
+- name: Install or Update flannel system container package
+  oc_atomic_container:
+    name: flanneld
+    image: "{{ flannel_system_image }}"
+    state: latest

--- a/ansible/roles/flannel/templates/flanneld.j2
+++ b/ansible/roles/flannel/templates/flanneld.j2
@@ -10,9 +10,9 @@ FLANNEL_ETCD_KEY="/{{ cluster_name }}/network"
 FLANNEL_ETCD_PREFIX="/{{ cluster_name }}/network"
 
 {% if etcd_url_scheme is defined and etcd_url_scheme == 'https' %}
-FLANNELD_ETCD_CAFILE="{{ flannel_etcd_ca_file }}"
-FLANNELD_ETCD_CERTFILE="{{ flannel_etcd_cert_file }}"
-FLANNELD_ETCD_KEYFILE="{{ flannel_etcd_key_file }}"
+FLANNEL_ETCD_CAFILE="{{ flannel_etcd_ca_file }}"
+FLANNEL_ETCD_CERTFILE="{{ flannel_etcd_cert_file }}"
+FLANNEL_ETCD_KEYFILE="{{ flannel_etcd_key_file }}"
 {% endif %}
 
 # Any additional options that you want to pass

--- a/ansible/roles/kubernetes-addons/files/common/kube-addons.sh
+++ b/ansible/roles/kubernetes-addons/files/common/kube-addons.sh
@@ -126,7 +126,7 @@ function create-resource-from-string() {
   local -r config_name=$4;
   local -r namespace=$5;
   while [ ${tries} -gt 0 ]; do
-    echo "${config_string}" | ${KUBECTL} --namespace="${namespace}" create -f - && \
+    echo "${config_string}" | ${KUBECTL} --namespace="${namespace}" apply -f - && \
         echo "== Successfully started ${config_name} in namespace ${namespace} at $(date -Is)" && \
         return 0;
     let tries=tries-1;

--- a/ansible/roles/master/defaults/main.yml
+++ b/ansible/roles/master/defaults/main.yml
@@ -53,3 +53,9 @@ kube_scheduler_options:
 kube_apiserver_additional_options: []
 kube_controller_manager_additional_options: []
 kube_scheduler_additional_options: []
+
+kube_apiserver_system_image: "{{ system_images_registry }}/{{ system_images_registry_namespace }}/kubernetes-apiserver:{{ system_images_tag }}"
+
+kube_controller_manager_system_image: "{{ system_images_registry }}/{{ system_images_registry_namespace }}/kubernetes-controller-manager:{{ system_images_tag }}"
+
+kube_scheduler_system_image: "{{ system_images_registry }}/{{ system_images_registry_namespace }}/kubernetes-scheduler:{{ system_images_tag }}"

--- a/ansible/roles/master/tasks/install.yml
+++ b/ansible/roles/master/tasks/install.yml
@@ -3,6 +3,10 @@
 - include: coreos.yml
   when: is_coreos
 
+- name: Install master system containers
+  include: system_container.yml
+  when: kube_use_master_system_container
+
 - include: ubuntu-14.04.yml
   when: ansible_distribution == 'Ubuntu' and ansible_distribution_major_version|int < 15
 

--- a/ansible/roles/master/tasks/system_container.yml
+++ b/ansible/roles/master/tasks/system_container.yml
@@ -1,0 +1,32 @@
+- name: Pull kubernetes-apiserver system container
+  command: atomic pull --storage=ostree {{ kube_apiserver_system_image }}
+  register: pull_result
+  changed_when: "'Pulling layer' in pull_result.stdout"
+
+- name: Pull kubernetes-controller-manager system container
+  command: atomic pull --storage=ostree {{ kube_controller_manager_system_image }}
+  register: pull_result
+  changed_when: "'Pulling layer' in pull_result.stdout"
+
+- name: Pull kubernetes-scheduler system container
+  command: atomic pull --storage=ostree {{ kube_scheduler_system_image }}
+  register: pull_result
+  changed_when: "'Pulling layer' in pull_result.stdout"
+
+- name: Install or Update kubernetes-apiserver system container package
+  oc_atomic_container:
+    name: kube-apiserver
+    image: "{{ kube_apiserver_system_image }}"
+    state: latest
+
+- name: Install or Update kubernetes-controller-manager system container package
+  oc_atomic_container:
+    name: kube-controller-manager
+    image: "{{ kube_controller_manager_system_image }}"
+    state: latest
+
+- name: Install or Update kubernetes-scheduler system container package
+  oc_atomic_container:
+    name: kube-scheduler
+    image: "{{ kube_scheduler_system_image }}"
+    state: latest

--- a/ansible/roles/node/defaults/main.yml
+++ b/ansible/roles/node/defaults/main.yml
@@ -21,7 +21,7 @@ kubelet_options:
   - "--cgroup-driver=systemd"
 
 kube_proxy_options:
-  - "--kubeconfig={{ kube_config_dir }}/proxy.kubeconfig"
+  - "--kubeconfig={{ kube_config_dir }}/proxy.kubeconfig --cluster-cidr={{ kube_service_addresses }}"
 
 kubelet_additional_options: []
 #  - "--host-network-sources=*"
@@ -30,3 +30,9 @@ kubelet_additional_options: []
 kube_proxy_additional_options: []
 
 disable_swap: true
+
+kubelet_system_image: "{{ system_images_registry }}/{{ system_images_registry_namespace }}/kubernetes-kubelet:{{ system_images_tag }}"
+
+kube_proxy_system_image: "{{ system_images_registry }}/{{ system_images_registry_namespace }}/kubernetes-proxy:{{ system_images_tag }}"
+
+

--- a/ansible/roles/node/tasks/install.yml
+++ b/ansible/roles/node/tasks/install.yml
@@ -3,6 +3,10 @@
 - include: coreos.yml
   when: is_coreos
 
+- name: Install node system containers
+  include: system_container.yml
+  when: kube_use_node_system_container
+
 - include: ubuntu-14.04.yml
   when: ansible_distribution == 'Ubuntu' and ansible_distribution_major_version|int < 15
 

--- a/ansible/roles/node/tasks/system_container.yml
+++ b/ansible/roles/node/tasks/system_container.yml
@@ -1,0 +1,21 @@
+- name: Pull kubernetes-kubelet system container
+  command: atomic pull --storage=ostree {{ kubelet_system_image }}
+  register: pull_result
+  changed_when: "'Pulling layer' in pull_result.stdout"
+
+- name: Pull kubernetes-proxy system container
+  command: atomic pull --storage=ostree {{ kube_proxy_system_image }}
+  register: pull_result
+  changed_when: "'Pulling layer' in pull_result.stdout"
+
+- name: Install or Update kubernetes-kubelet system container package
+  oc_atomic_container:
+    name: kubelet
+    image: "{{ kubelet_system_image }}"
+    state: latest
+
+- name: Install or Update kubernetes-proxy system container package
+  oc_atomic_container:
+    name: kube-proxy
+    image: "{{ kube_proxy_system_image }}"
+    state: latest

--- a/ansible/scripts/ansible.cfg
+++ b/ansible/scripts/ansible.cfg
@@ -1,2 +1,3 @@
 [defaults]
 roles_path = ../roles
+library = ../library

